internal/dex/code_verifier.go

```diff
@@ -0,0 +1,39 @@
+package dex
+
+import (
+	"crypto/sha256"
+	"encoding/base64"
+	"strings"
+
+	"github.com/dchest/uniuri"
+)
+
+// CodeVerifier is an helper struct used to create a code_challenge for the PKCE
+// Ref: https://www.oauth.com/oauth2-servers/pkce/
+type CodeVerifier struct {
+	Value string
+}
+
+// NewCodeVerifier returns a cryptographic secure random CodeVerifier of a fixed length (32)
+func NewCodeVerifier() *CodeVerifier {
+	return NewCodeVerifierWithLen(32)
+}
+
+// NewCodeVerifier returns a cryptographic secure random CodeVerifier of the specified length
+func NewCodeVerifierWithLen(len int) *CodeVerifier {
+	return &CodeVerifier{Value: uniuri.NewLen(len)}
+}
+
+// ChallengeS256 returns an encoded SHA256 code_challenge of the code_verifier
+func (c *CodeVerifier) ChallengeS256() string {
+	h := sha256.New()
+	h.Write([]byte(c.Value))
+	hash := h.Sum(nil)
+
+	encoded := base64.StdEncoding.EncodeToString(hash)
+	encoded = strings.Replace(encoded, "+", "-", -1)
+	encoded = strings.Replace(encoded, "/", "_", -1)
+	encoded = strings.Replace(encoded, "=", "", -1)
```

Comment:
What's with the replacements?

Reply:
That is part of the RFC. Basically the base64 needs to be "url safe", and so it's actually a `BASE64URL-ENCODE`, not just a `BASE64-ENCODE`.

https://datatracker.ietf.org/doc/html/rfc7636#section-4.2
example: https://www.oauth.com/oauth2-servers/pkce/authorization-request/

Reply:
I think you can use [`RawURLEncoding`](https://cs.opensource.google/go/go/+/master:src/encoding/base64/base64.go;l=128) instead of `StdEncoding` above and achieve the same thing.

Reply:
Ah! That's interesting, thanks! Yes, it seems what we need: https://cs.opensource.google/go/go/+/master:src/encoding/base64/base64.go;l=53;drc=bd8a5b00fcf71fb711ba8996a880b5b07f7b3634